src/core/connection/SSDConverter.hpp

```diff
@@ -0,0 +1,211 @@
+        return std::make_pair(std::nullopt, logList);
+    }
+
+    // casting to object
+    QJsonObject obj = document.object();
+    //
+    // obj.airport
+    MUST_STRING("airport");
+    const QString airport = obj["airport"].toString();
+    // obj.port
+    MUST_PORT("port");
+    const int port = obj["port"].toInt();
+    // obj.encryption
+    MUST_STRING("encryption");
+    const QString encryption = obj["encryption"].toString();
+
+    // check: rc4-md5 is not supported by v2ray-core
+    // TODO: more checks, including all algorithms
+    if (encryption == "rc4-md5") {
+        logList << QObject::tr("invalid ssd link: rc4-md5 encryption is not supported by v2ray-core");
+        return std::make_pair(std::nullopt, logList);
+    }
+
+    // obj.password
+    MUST_STRING("password");
+    const QString password = obj["password"].toString();
+    // obj.servers
+    MUST_ARRAY("servers");
+    //
+    QList<std::pair<QString, ShadowSocksServerObject>> serverList;
+    //
+
+    // iterate through the servers
+    for (QJsonValueRef server : obj["servers"].toArray()) {
+        SERVER_SHOULD_BE_OBJECT(server);
+        QJsonObject serverObject = server.toObject();
+        ShadowSocksServerObject ssObject;
+        // encryption
+        ssObject.method = encryption;
+        // password
+        ssObject.password = password;
+        // address :-> "server"
+        SHOULD_STRING("server");
+        const QString serverAddress = serverObject["server"].toString();
+        ssObject.address = serverAddress;
+
+        // port selection:
+        //   normal:     use global settings
+        //   overriding: use current config
+        if (serverObject["port"].isUndefined()) {
+            ssObject.port = port;
+        } else if (auto currPort = serverObject["port"].toInt(-1); port >= 0 && port <= 65535) {
+            ssObject.port = currPort;
+        } else {
+            logList << QObject::tr("warning: invalid port encountered. using fallback value.");
+            ssObject.port = port;
+        }
+
+        // name decision:
+        //   untitled: using server:port as name
+        //   entitled: using given name
+        QString nodeName;
+
+        if (serverObject["remarks"].isUndefined()) {
+            nodeName = QString("%1:%2").arg(ssObject.address).arg(ssObject.port);
+        } else if (serverObject["remarks"].isString()) {
+            nodeName = serverObject["remarks"].toString();
+        } else {
+            logList << QObject::tr("warning: invalid name encountered. using fallback value.");
+            nodeName = QString("%1:%2").arg(ssObject.address).arg(ssObject.port);
+        }
+
+        // ratio decision:
+        //   unspecified: ratio = 1
+        //   specified:   use given value
+        double ratio = 1.0;
+
+        if (auto currRatio = serverObject["ratio"].toDouble(-1.0); currRatio != -1.0) {
+            ratio = currRatio;
+        } else if (!serverObject["ratio"].isUndefined()) {
```

Comment:
Here, you firstly access the `serverObject` with `operator[]` in the first `if` branch, then check for undefined in the other branch.

This will explain everything.
![image](https://user-images.githubusercontent.com/59914293/75089982-3e315880-5599-11ea-8e8c-5893975587ee.png)

Finally, you will get `x = false`, because the key "key" is set to `Null` internally by QJsonObject when calling `toDouble`

To check it carefully, you may want to use `jsonObj.contains("")` first instead of using `isUndefined`.

See: https://github.com/Qv2ray/Qv2ray/issues/129 https://github.com/Qv2ray/Qv2ray/blob/1074ac0882b3913945611efa29d06984fea4ff2a/src/ui/w_MainWindow.cpp#L428 and  https://github.com/Qv2ray/Qv2ray/commit/aec993711d06cb9d9830707a967d7035b63af7be